### DISCUSSION TITLE
Possible to have multi value ratings (1-5 or 1 - 10) ?

### DISCUSSION THREAD

**Discussion by mbhangui** [jcorporation/myMPD#565] [Ideas]
First let me thank you for forking ympd and making it better and adding the mpd sticker functionality. You have done a marvellous job. You also have clarity on what myMPD should become. You are maintaining play counts, skip counts, last played and also have smart playlists. Most of these feature are just not available anywhere. The only other client that I came across which uses mpd sticker is cantata. But cantata can be run only under a desktop (gnome, kde, mate, etc). Cantata uses 1-10 point rating for songs. Half star means 1, a full star means 2 and one has 5 stars on the UI that one can click to rate a song from 1 to 10. But it does not maintain play counts, last played, etc

So I wrote a daemon [mpdev](https://github.com/mbhangui/mpdev) which connects to the mpd socket on 6600 and uses the IDLE comment to listen to mpd events and execute scripts based on the event. This is something which, you are also doing. Using mpdev, I'm constantly updating the play counts, skip counts, scrobbling to last.fm, libre.fm, even if I play my player in headless mode. Over a period of time I have built a database and use a shell script to generate smart playlist to generate songs that I love and like. The only issue that I had was a player that could rate songs. On a Desktop that is cantata, but on android, ios I was looking for something that would be running as a webserver or under a webserver. This is where I find that myMPD satisfies all my needs. The only functionality that I miss is the multi-point rating system. You have 0 for thumbs down, 1 for neutral and 2 for thumbs up. Can we have a feature where repeatedly pressing thumbs up will increment rating by +1 till it reaches a max configured value. Similarly thumbs down can decrement by -1 till the rating reaches zero? The other option is to have 5 stars in the UI. It would be awesome to have have something like this, but I'm not sure if it will fit into your scheme of things. I really love the UI and the simplicity of not installing a webserver like apache, nginx, etc.

**Comment by jcorporation** [comment]
I think this is worth a discussion. 

I think thumb up / down and a star rating should not be merged. Both rating methods have there advantages. 

Thumbs can be mapped to love/hate from listenbrainz.

Stars are more granular. But Half-Star ratings seems to me too fine granular and how should they be implemented for the user interface?

Any opinions from other myMPD users?

**Reply by john-** [reply]
The other option @mbhangui mentioned was having logic based on repeatedly selecting thumbs up/down.  This could still be mapped to musicbrainz (a value greater than 1 is equal to "love").

A tooltip can display current number of thumbs up/down to leave UI basically unchanged.  Alternately, add the count to the icon.   Or the hand icon can get progressively greener based on how liked (or disliked) it is :)

In the end I feel the current implementation works for me because whether I like something is pretty binary (I either like it or I don't).  However, depending on my mood I will not listen to certain things even if I like them.   That said associating the favorite with how I feel at the time is a different feature :)

**Comment by jcorporation** [comment]
> A tooltip can display current number of thumbs up/down to leave UI basically unchanged. Alternately, add the count to the icon.

Good idea!

**Comment by bartkl** [comment]
To add my two cents, I would really love to rate *albums* using a five-point scale. For tracks, giving thumbs up/down seems to play nicely with that as well. At least I definitely think it's wise to not replace the thumbs up/down system if people are enjoying that :).

Regarding the half stars question: I wouldn't need them, but many people seem to like this level of granularity. Perhaps better to be on the safe side and support it, perhaps making it possible through some config setting.

Would love to hear if you're considering implementing ratings at the album level this way.

Thanks a lot for this great player. I'm amazed by its speed on my Raspberry Pi and great features!

**Reply by jcorporation** [reply]
MPD 0.24 added the ability to attach stickers to more object. Adding a rating feature for albums and for playlists are now possible.

#1134

**Reply by bartkl** [reply]
That woud be amazing for me. Is such a feature something you would consider implementing? Would you perhaps like me to create a separate discussion or issue for that?

**Reply by jcorporation** [reply]
It is already on the roadmap: #1134

**Reply by bartkl** [reply]
Great, thanks!

**Comment by jcorporation** [comment]
> Would love to hear if you're considering implementing ratings at the album level this way.

Ratings are always at song level, because MPD support stickers only for songs.

I personally have no big preference for stars or thumbs rating.

**Comment by jcorporation** [comment]
@mbhangui: which sticker names does your mpdev client use? This will be interesting for this project: https://github.com/jcorporation/mpd-stickers

**Reply by mbhangui** [reply]
This is a standard sqlite db that mpd creates when you have `sticker_file "/var/lib/mpd/sticker.sql"` in mpd.conf for example. The schema is as below. The same schema is used by cantata music player to record ratings. The cantata client simply uses the `value` field to insert a number from 0-10 indicating rating and displays the same in a 5 point scale (5 stars means 10, 4 stars/asterix means 8). My mpdev client simply reuses the values in the database as created by cantata client. The 4 fields that mpd creates for the sticker database are type, uri, name, value. If you want to use your own fields, let me know. I can adjust things at my end and also update cantata player to use the same fields that myMPD uses.

```
$ sqlite3 sticker.sql
sqlite> .schema
sqlite3 sticker.sql
SQLite version 3.40.1 2022-12-28 14:03:47
Enter ".help" for usage hints.
sqlite> .schema
CREATE TABLE sticker(  type VARCHAR NOT NULL,   uri VARCHAR NOT NULL,   name VARCHAR NOT NULL,   value VARCHAR NOT NULL);
CREATE UNIQUE INDEX sticker_value ON sticker(type, uri, name);
```

Few records for which I have used cantata client to update the ratings are as below.

```
sqlite> select * from sticker limit 5;
song|ABBA/ABBA - Flac/01 Mamma mia.flac|rating|6
song|ABBA/ABBA - Flac/02 Hey hey helen.flac|rating|6
song|ABBA/ABBA - Flac/03 Tropical loveland.flac|rating|5
song|ABBA/ABBA - Flac/04 Sos.flac|rating|6
song|ABBA/ABBA - Flac/05 Man in the middle.flac|rating|4
```

**Reply by jcorporation** [reply]
Schema is always the same across all mpd clients. At interest are the names and the possible values of the stickers, rating, etc. could you add your used sticker names to the documentation?

**Reply by mbhangui** [reply]
Is this ok for you? This is what Cantata uses and hence mpdev also uses the same

Sticker DB Field Name|Format|Value
--------------------------------|-----------|--------
type|text|song
uri|text|uri of the song
name|text|rating
value|number|0-10

**Reply by jcorporation** [reply]
Field names are unnecessary as the schema of the stickers database is fixed and defined by MPD. Only the used sticker names with possible values and a description is interesting. Look at https://github.com/jcorporation/mpd-stickers for examples, how it should be documented. A pull request for this repo will be great. Thanks for your effort!

**Reply by mbhangui** [reply]
There is no need to raise a PR. You have already documented the ratiing sticker that cantata uses accurately (0-10 int values). Once you have the rating scheme in myMPD, I will modify my mpdev to use your database for playcounts, skipped count, last played, etc. Currently, I'm maintaining a separate db which mpdev populates by listening for mpd events. Since you already have that, there is no point in duplicating data. I will also move out of using cantata on my macbook and linux laptops and use myMPD for all listening/playback/control of mpd.

Personally, I'm excited by this project of yours.
